lib/router/index.ts

```diff
@@ -0,0 +1,222 @@
+
+// substitute calls _doSubstitute on each property of `obj`.
+function substitute(obj, indicator, subber) {
+  return _.mapObject(obj, (val) => _doSubstitute(val, indicator, subber));
+}
+
+function deepLookupPath(obj, path) {
+  if (path.length === 0 || obj[path[0]] === undefined) {
+    return undefined;
+  }
+  if (path.length > 1) {
+    return deepLookupPath(obj[path[0]], path.slice(1));
+  }
+  return obj[path[0]];
+}
+
+function deepLookup(obj, field) {
+  return deepLookupPath(obj, field.split("."));
+}
+
+function fieldMatches(obj, field, values) {
+  const val = deepLookup(obj, field);
+  return _.any(values, (poss) => val === poss);
+}
+
+class Rule {
+  name = null;
+  matchers = null;
+  output = null;
+
+  constructor(name, matchers, output) {
+    this.name = name;
+    this.matchers = matchers;
+
+    const envMissing = [];
+    this.output = substitute(output, "\\$", (k) => {
+      const val = process.env[k];
+      if (val == null) {
+        envMissing.push(k);
+      }
+      return val;
+    });
+    if (envMissing.length > 0) {
+      throw new Error(`Missing env var(s): ${envMissing.join(", ")}`);
+    }
+    this.output.rule = this.name;
+  }
+
+  // matches returns true if `msg` matches against this rule
+  matches(msg) {
+    const matches = _.mapObject(this.matchers, (values, field) => fieldMatches(msg, field, values));
+    return _.all(matches);
+  }
+
+  // returns the output with kv substitutions performed
+  outputFor(msg) {
+    const flatMsg = _.deepToFlat(msg);
+    return substitute(this.output, "%", (k) => flatMsg[k] || "KEY_NOT_FOUND");
+  }
+}
+
+// transformValidation transforms a result from jsonschema.validate to prefix
+// all errors with `name`.
+function transformValidation(res, name) {
+  return {
+    valid: res.valid,
+    errors: res.errors.map((err) => `${name}: ${err.stack}`),
+  };
+}
+
+// validateRoutes ensures that `routes` matches the config schema. We have this
+// function instead of just doing a plain jsonschema.validate in order to get
+// better error messages for the "output" object (by default jsonschema would
+// just tell you that the output block doesn't match any of the known output
+// formats, but won't tell you what's wrong because it doesn't let you
+// condition on the output.type property).
+function validateRoutes(routes) {
+  const validator = new jsonschema.Validator();
+  validator.addSchema({definitions: schemaDefs}, "/config");
+  const results = _.mapObject(routes, (rule, name) => {
+    // validate that the rule has the appropriate structure
+    const structureVal = validator.validate(rule, {
+      type: "object",
+      additionalProperties: false,
+      required: ["matchers", "output"],
+      properties: {
+        matchers: {
+          type: "object",
+        },
+        output: {
+          type: "object",
+          properties: {
+            type: {
+              type: "string",
+              enum: ["metrics", "alert", "analytics", "notification"],
```

Comment:
can we make sure these are all singular or all plural?

I'd vote for all plural.
```
"metrics", 
"alerts", 
"analytics", 
"notifications"
```